### PR TITLE
[SPARK-29493][SQL] Arrow MapType support

### PR DESCRIPTION
**Pull request by jalpan-randeri**
### What changes were proposed in this pull request?
This pull request add support for Arrow MapType into Spark SQL.

### Why are the changes needed?
Without this change User's of spark are not able to query data in spark if one of columns is stored as map and Apache Arrow execution mode is preferred by user. 
More info: https://issues.apache.org/jira/projects/SPARK/issues/SPARK-29493

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Introduced few unit tests around map type in existing arrow test suit

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@jalpan-randeri can you rebase? Arrow 0.15.1 upgrade is merged now.

**Comment by jalpan-randeri** [comment]
@HyukjinKwon  Done.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
cc @BryanCutler

**Comment by BryanCutler** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala]: `valueContainsNull` is for map values have null or not. For FieldType here, I think it is like ArrayType and StructType cases, we should use given nullable parameter.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala]: please follow usual indent style.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala]: can't we set null for a map vector?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala
Comment (jalpan-randeri) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala]: Nice catch. Fixed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala
Comment (jalpan-randeri) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala
Comment (jalpan-randeri) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala]: map vector implemented as an extension of list vector. it is represented as list of struct. it seems we can not set key as null. I found following comments in code,  
```
if a map is set at a given index, there must be an entry. In other words, the StructVector data is non-nullable. 
```

- https://github.com/apache/arrow/blob/master/java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java#L36

**Comment by maropu** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala]: Can we keep this test by using another type? (e.g., CalendarIntervalType or UDT).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala]: We might want to use `while` loop to be performant like `ArrayWriter`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala
Comment (jalpan-randeri) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala]: Done, modified this test to work with CalendarInternvalType

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala
Comment (jalpan-randeri) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala]: Done. Yes, i was using incorrect constructor

File: sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowWriterSuite.scala
Comment (jalpan-randeri) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowWriterSuite.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala
Comment (jalpan-randeri) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala
Comment (jalpan-randeri) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala]: yup. i mistakenly reverted this commit, fixed in new revision

Comment (ueshin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala]: Seems not yet fixed?

I guess @viirya is asking:

```scala
val mapType = new FieldType(nullable, new ArrowType.Map(false), null)
```

, and I agree with it.

WDYT?

**Comment by jalpan-randeri** [comment]
jenkins, retest please

**Comment by maropu** [comment]
@jalpan-randeri Thanks for your work. For better commit logs, can you use the new PR template in the description above? https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE

**Comment by jalpan-randeri** [comment]
Sure. Shall I squash my changes into one commit and use the template you mentioned.

**Comment by maropu** [comment]
Thanks, @jalpan-randeri

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala
Comment (jalpan-randeri) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala]: Oh, i understood now, my bad i was getting confused with other fields.

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestData.scala
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestData.scala]: Changes in this file are not related to this PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala]: Do we need this?
Seems like complex types are handled in `toArrowField` and `fromArrowField`?

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestData.scala
Comment (jalpan-randeri) [reply sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestData.scala]: Fixed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala
Comment (jalpan-randeri) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala]: Fixed

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestData.scala
Comment (ueshin) [reply sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestData.scala]: Could you revert the other changes in the file?

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.
Could you remove `[WIP]` from the PR title?

**Comment by maropu** [comment]
The code itself looks good, but we need to update some docs, too (follow-up?);
https://github.com/apache/spark/blob/d7843dde0f82551f0481885feb15acd63dd554c0/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L1543
https://github.com/apache/spark/blob/master/docs/sql-pyspark-pandas-with-arrow.md#supported-sql-types

**Comment by jalpan-randeri** [comment]
Done. Updated docs

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: It seems we already support BinaryType? I saw @BryanCutler updating arrow to v0.15.1.

**Review by maropu — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (jalpan-randeri) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Fixed

**Comment by maropu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: SparkR Arrow optimization still won't work due to https://github.com/apache/spark/blob/93ea353cae0ccc187028f79024b239893ebaeb96/R/pkg/R/types.R#L87-L120

You can revert this change back. Let's do this separately in a separate PR.

**Review by HyukjinKwon — APPROVED** [review]
Looks fine to me too except one comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (ueshin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: +1
Shall we revert the last two commits related to docs for now?

File: docs/sql-pyspark-pandas-with-arrow.md
Comment (BryanCutler) [inline docs/sql-pyspark-pandas-with-arrow.md]: MapType still cannot be supported for pandas udfs because Maps have only recently been added to pyarrow and will only be available in pyarrow 1.0, see ARROW-6904. Could you revert this?

**Review by BryanCutler — COMMENTED** [review]
Thanks for doing this @jalpan-randeri ! I haven't had a chance to look through it yet, but I just want to make sure that it is clear that this is only enabling the Java side and for PySpark MapType is still unsupported and gives the user a good error message. We won't be able to support MapTypes for pandas udfs until pyarrow 1.0.

File: docs/sql-pyspark-pandas-with-arrow.md
Comment (maropu) [reply docs/sql-pyspark-pandas-with-arrow.md]: Ah, I see. Thanks for the check.

Comment (jalpan-randeri) [reply docs/sql-pyspark-pandas-with-arrow.md]: Done. Reverted last 2 commits

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (jalpan-randeri) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Done

**Comment by maropu** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by jalpan-randeri** [comment]
wow!
@BryanCutler  Thank you for giving me opportunity to work on this. 
@HyukjinKwon  @ueshin  @maropu @viirya Thank you for reviewing my work. 
I hope to contribute more in future.

**Comment by HyukjinKwon** [comment]
This was a pretty important change so it took a while until it gets merged. Thanks for your work and patience :-).

**Comment by gatorsmile** [comment]
@jalpan-randeri What is the Apache Arrow execution mode ?

**Comment by jalpan-randeri** [comment]
it is just setting this property.
``
In [7]: spark.conf.set("spark.sql.execution.arrow.enabled", True)
``

**Pull request closed without merging**
